pvlib/iotools/psm3.py

```diff
@@ -0,0 +1,133 @@
+"""
+Get PSM3 TMY
+see https://developer.nrel.gov/docs/solar/nsrdb/psm3_data_download/
+"""
+
+import io
+import csv
+import datetime
+import requests
+import pandas as pd
+import pytz
+
+URL = "http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv"
+
+# 'relative_humidity', 'total_precipitable_water' are note available
+ATTRIBUTES = [
+    'air_temperature', 'dew_point', 'dhi', 'dni', 'ghi', 'surface_albedo',
+    'surface_pressure', 'wind_direction', 'wind_speed']
+
+
+def get_psm3(latitude, longitude, tmy='tmy', interval=60):
+    """
+    Get PSM3 data
+
+    Parameters
+    ----------
+    latitude : float or int
+        in decimal degrees, between -90 and 90, north is positive
+    longitude : float or int
+        in decimal degrees, between -180 and 180, east is positive
+    tmy : str
+        PSM3 parameter "name", see notes below for options, default ``'tmy'``
+    interval : integer
+        interval size in minutes, can be only either 30 or 60, default 60
+
+    Returns
+    -------
+    headers : dict
+        meta data from NREL PSM-3 about the record, see notes for fields
+    data : pandas.DataFrame
+        timeseries data from NREL PSM3
+
+    Raises
+    ------
+    requests.HTTPError
+
+    Notes
+    -----
+    The ``tmy`` parameter 'names' must be a single value from the following
+    list:
+
+        ['1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005',
+         '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
+         '2014', '2015', '2016', '2017', 'tmy', 'tmy-2016', 'tmy-2017',
+         'tdy-2017', 'tgy-2017']
+
+    The return is a tuple with two items. The first item is a header with meta
+    data from NREL PSM3 about the record containing the following fields:
+
+    * Source
+    * Location ID
+    * City
+    * State
+    * Country
+    * Latitude
+    * Longitude
+    * Time Zone
+    * Elevation
+    * Local Time Zone
+    * Dew Point Units
+    * DHI Units
+    * DNI Units
+    * GHI Units
+    * Temperature Units
+    * Pressure Units
+    * Wind Direction Units
+    * Wind Speed
+    * Surface Albedo Units
+    * Version
+
+    See Also
+    --------
+    pvlib.iotools.read_tmy2, pvlib.iotools.read_tmy3
+
+    References
+    ----------
+    [1] `NREL Developer Network - Physical Solar Model (PSM) v3
+        <https://developer.nrel.gov/docs/solar/nsrdb/psm3_data_download/>`_
+    [2] `NREL National Solar Radiation Database (NSRDB)
+        <https://nsrdb.nrel.gov/>`_
+    """
+    longitude = ('%9.4f' % longitude).strip()
+    latitude = ('%8.4f' % latitude).strip()
+    params = {
```

Comment:
Do we need to provide a way for users to specify at least their own api key? I suspect so. I'd hope that the name, email, affiliation, reason fields are not required and are linked to the api key when a user registers.

Reply:
@cdeline do you know what NREL's stance is on this?

Reply:
I'll double-check with Manajit.  From the API documentation [https://nsrdb.nrel.gov/api-instructions](url) there are several things requested in addition to api_key. Unclear if these are stored at time of API key registration :


`your_name = 'John+Smith'  # Your full name, use '+' instead of spaces.`
`reason_for_use = 'beta+testing' # Your reason for using the NSRDB.`
`your_affiliation = 'my+institution' # Your affiliation`
`your_email = 'john.smith@server.com' # Your email address`
`mailing_list = 'true' # Please join our mailing list so we can keep you up-to-date on new developments.`

Reply:
@mikofski your timing with this PR could not have been better.  FYI, `get_psm3` works passing my own API key and no other meta data.

Reply:
Here's a response from the NSRDB API developer, Paul Edwards:

>OK, as to the first question about user metadata, technically speaking the email address is the only field that is strictly required because the system uses this to automatically communicate messages back to the user if necessary (such as a download failure etc). However we would strongly encourage your team to include the rest! We use the other fields to collect user stats for internal use and these stats have been vital in our ability to report accurately back to DOE on the usefulness of these services. Also, we don’t have access to the fields associated with the API key. In fact when developer.nrel.gov proxies the request through to our code it doesn’t even include the API key at all. 

>Regarding the “DEMO_KEY” api key, that key is severely rate limited, and is also used in tons of examples all over the place for 100s of endpoints. Consequently it is not uncommon to find that using that key will result in rejected requests due to the rate limit. My recommendation is to go ahead and include it in your examples etc, but with strong disclaimers along the lines of what I just said… nothing is going to work right until one uses their own API key. There is actually a documentation page with more details that your team could use to inform how to write code to report back to users when the rate limit is the cause of failures if that is something you’re interested in.

https://developer.nrel.gov/docs/rate-limits/


>Oh yeah, I missed one of your points. Yes, I think it would be neat if the “reason” attribute was something like “pvlib”. We could even pull stats for the PVLib team after a while if that was useful to them.

>Let me know if this helps, 
Paul

Reply:
So no, it doesn't appear that user information is connected back to the api key for any reason.

Here's what I would recommend based on this feedback:
1. Include 'pvlib' somewhere in the api call, like 'reason':'pvlib'.  So we can track the number of users that take advantage of this function.
2. Have users pass their email address as a separate function input along with the api key.
3. Include in the documentation a suggestion that users optionally hard-code other specific details such as affiliation.  
4. Set mailing_list = false by default if we're requesting email addresses (?)

Reply:
Thanks @cdeline. 

Maybe arg `api_key` and kwargs for the other parameters (full_name, email, affiliation, reason, mailing_list). Use the demo key in examples/tests. I doubt many people will use their own API key if it's only a kwarg.